File: 0022.generate_parentheses/solution.py
from typing import List


class Solution:
    def generate_parenthesis(self, n: int) -> List[str]:
        def backtrack(s="", left=0, right=0):
            if len(s) == 2 * n:
                ans.append(s)
                return
            if left < n:
                backtrack(s + "(", left + 1, right)
            if right < left:
                backtrack(s + ")", left, right + 1)

        ans = []
        backtrack()
        return ans


if __name__ == "__main__":
    assert len(Solution().generate_parenthesis(2)) == 2
    assert len(Solution().generate_parenthesis(3)) == 5

File: 0022.generate_parentheses/solution.go
package leetcode0022

func generateParenthesis(n int) []string {
	ans := make([]string, 0)
	backtrack("", n, n, &ans)
	return ans
}

func backtrack(s string, left, right int, ans *[]string) {
	if left == 0 { // s中没左括号了，其余的都是不匹配的值
		for i := 0; i < right; i++ {
			s += ")"
		}
		*ans = append(*ans, s)
		return
	}
	backtrack(s+"(", left-1, right, ans)
	if left < right {
		backtrack(s+")", left, right-1, ans)
	}
}

func main() {
	generateParenthesis(3)
}
